### ISSUE TITLE
[RFC] Allow context initialization to use `ocaml` binaries from workspace

### ISSUE THREAD

**Issue by ejgallego** [ocaml/dune#4433]
It would be very cool if we could compose the OCaml compiler itself in a Dune workspace (c.f. https://github.com/ocaml/ocaml/issues/10315 ). A key step is to actually teach Dune to also look for OCaml binaries produced by rules in the current workspace.

This is subtle, thus I'm making an issue to discuss possible approaches; as of today, the code initializing the context does:
```ocaml
    let* ocamlc =
      get_tool_using_findlib_config "ocamlc" >>= function
      | Some x -> Memo.Build.return x
      | None -> (
        which "ocamlc" >>| function
        | Some x -> x
        | None -> prog_not_found_in_path "ocamlc")
    in
    let dir = Path.parent_exn ocamlc in
    let get_ocaml_tool prog =
      get_tool_using_findlib_config prog >>| function
      | Some x -> Ok x
      | None -> (
        match Program.best_path dir prog with
        | Some p -> Ok p
        | None ->
          let hint =
            sprintf "ocamlc found in %s, but %s/%s doesn't exist (context: %s)"
              (Path.to_string dir) (Path.to_string dir) prog
              (Context_name.to_string name)
          in
          Error
            (Action.Prog.Not_found.create ~context:name ~program:prog ~loc:None
               ~hint ()))
    in
```
thus, findlib is tried first, if not, path is tried (IIUC)

I could image two ways to tweak this:
- make ocaml binary resolution just use the regular mechanism from `Super_context.resolve_program`
- add yet a first special case for the resolution on the current workspace

Good news is that all the tools are already `Action.Prog.t`  in `Context.t` , except for `ocaml` which is a path, I wonder why.

**Comment by jeremiedimino** [comment]
> make ocaml binary resolution just use the regular mechanism from Super_context.resolve_program

Are you thinking to use `Super_context.resolve_program` inside `context.ml`? Or rather to remove the `ocamlc`, `ocamlopt` fields from `context.t` and use `Super_context.resolve_program` whenever we need to access `ocamlc` or `ocamlopt`?

**Comment by nojb** [comment]
Just trying to understand what is involved in this proposal (which I support!): assuming that `dune` can run `ocamlc` and `ocamlopt` built as part of the workspace, how will it work with respect to the standard library? Right now, the OCaml toolchain hard-codes the *installation* path to the standard library in the compiler drivers.

Is the idea that `dune` will detect that the compilers are being built as part of the current workspace and somehow force the stdlib path (eg by using `OCAMLLIB` environment variable)?

**Comment by ejgallego** [comment]
> Is the idea that `dune` will detect that the compilers are being built as part of the current workspace and somehow force the stdlib path (eg by using `OCAMLLIB` environment variable)?

Indeed, there will some plumbing needed to get a fully working setup, I think on both dune and the OCaml side, hopefully it should be just a couple of flags.

In Coq for example we have to signal that a certain library is special (the prelude) by using the `(boot)`  flag in their `(coq.theory ...)` definition.
